Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/main.kt

```diff
@@ -329,27 +329,42 @@ private fun processCLib(args: Array<String>, additionalArgs: Map<String, Any> =
         }
     }
 
+    val compiledFiles = compileSources(nativeLibsDir, tool, cinteropArguments)
+
     return when (stubIrOutput) {
         is StubIrDriver.Result.SourceCode -> {
             argsToCompiler(staticLibraries, libraryPaths)
         }
         is StubIrDriver.Result.Metadata -> {
-            val args = LibraryCreationArguments(
+            createInteropLibrary(
                     metadata = stubIrOutput.metadata,
-                    nativeBitcodePath = nativeOutputPath,
+                    nativeBitcodeFiles = compiledFiles + nativeOutputPath,
                     target = tool.target,
                     moduleName = moduleName,
                     outputPath = cinteropArguments.output,
                     manifest = def.manifestAddendProperties,
                     dependencies = allLibraryDependencies,
                     nopack = cinteropArguments.nopack
             )
-            createInteropLibrary(args)
             return null
         }
     }
 }
 
+private fun compileSources(
+        nativeLibsDir: String,
+        toolConfig: ToolConfig,
+        cinteropArguments: CInteropArguments
+): List<String> = cinteropArguments.compileSource.map { source ->
+    // Mangle path into file name to avoid collisions.
+    val mangledFileName = File(source).invariantSeparatorsPath.replace('/', '_')
```

Comment:
Not sure if this is ok when absolute paths involved.
It would be much simpler to use something like `$index_$shortName.bc`.